File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = list()
        match = {')': '(',  ']': '[', '}': '{'}
        for c in s:
            if c not in match.keys():  # open parenthesis
                stack.append(c)
            else:  # closed parenthesis
                # if there isn't any open parenthesis or parenthesis does not match
                if not stack or stack[-1] != match[c]:
                    return False
                stack.pop()
        
        return False if stack else True
        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> open;
        unordered_map<char, char> match = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        
        for (const auto &c: s) {
            // opened parenthesis
            if (match.find(c) == match.end()) {
                open.push(c);
            } else {
                if (open.empty()) return false;  // opened parenthesis does not exist
                if (open.top() != match[c]) return false;  // parenthesis does not match
                open.pop();
            }
        }
        return open.empty();
    }
};